thor/clusters.py

```diff
@@ -356,3 +400,319 @@ def _find_clusters_dbscan(points, eps, min_samples):
+
+    Returns
+    -------
+    list
+        If clusters are found, will return a list of numpy arrays containing the
+        observation IDs for each cluster. If no clusters are found, will return np.NaN.
+    """
+    logger.debug(f"cluster: vx={vx} vy={vy} n_obs={len(obs_ids)}")
+    xx = x - vx * dt
+    yy = y - vy * dt
+
+    X = np.stack((xx, yy), 1)
+
+    clusters = find_clusters(X, radius, min_obs, alg=alg)
+    clusters, arc_lengths = filter_clusters_by_length(
+        clusters,
+        dt,
+        min_obs,
+        min_arc_length,
+    )
+
+    if len(clusters) == 0:
+        return Clusters.empty(), ClusterMembers.empty()
+    else:
+
+        cluster_ids = []
+        cluster_num_obs = []
+        cluster_members_cluster_ids = []
+        cluster_members_obs_ids = []
+        for cluster in clusters:
+            id = uuid.uuid4().hex
+            obs_ids_i = obs_ids[cluster]
+            num_obs = len(obs_ids_i)
+
+            cluster_ids.append(id)
+            cluster_num_obs.append(num_obs)
+            cluster_members_cluster_ids.append(np.full(num_obs, id))
+            cluster_members_obs_ids.append(obs_ids_i)
+
+        clusters = Clusters.from_kwargs(
+            cluster_id=cluster_ids,
+            vtheta_x=np.full(len(cluster_ids), vx),
+            vtheta_y=np.full(len(cluster_ids), vy),
+            arc_length=arc_lengths,
+            num_obs=cluster_num_obs,
+        )
+
+        cluster_members = ClusterMembers.from_kwargs(
+            cluster_id=np.concatenate(cluster_members_cluster_ids),
+            obs_id=np.concatenate(cluster_members_obs_ids),
+        )
+
+    return clusters, cluster_members
+
+
+def cluster_velocity_worker(
+    vx: npt.NDArray[np.float64],
+    vy: npt.NDArray[np.float64],
+    obs_ids: npt.ArrayLike,
+    x: npt.NDArray[np.float64],
+    y: npt.NDArray[np.float64],
+    dt: npt.NDArray[np.float64],
+    radius: float = 0.005,
+    min_obs: int = 5,
+    min_arc_length: float = 1.0,
+    alg: Literal["hotspot_2d", "dbscan"] = "dbscan",
+) -> Tuple[Clusters, ClusterMembers]:
+    """
+    Helper function for parallelizing cluster_velocity. This function takes a
+    batch or chunk of velocities and returns the clusters and cluster members
+    for that batch.
+
+    """
+    clusters_list = []
+    cluster_members_list = []
+    for vx_i, vy_i in zip(vx, vy):
+        clusters_i, cluster_members_i = cluster_velocity(
+            obs_ids,
+            x,
+            y,
+            dt,
+            vx_i,
+            vy_i,
+            radius=radius,
+            min_obs=min_obs,
+            min_arc_length=min_arc_length,
+            alg=alg,
+        )
+        clusters_list.append(clusters_i)
+        cluster_members_list.append(cluster_members_i)
+
+    return qv.concatenate(clusters_list), qv.concatenate(cluster_members_list)
+
+
+cluster_velocity_remote = ray.remote(cluster_velocity_worker)
+cluster_velocity_remote.options(
+    num_returns=1,
+    num_cpus=1,
```

Comment:
For `num_returns=1`, I found it better to explicitly tell ray that I'm returned a tuple of two objects rather than two objects (`num_returns=2`). I couldn't figure out the pattern to take two object references from a single function with `ray.wait` and then `ray.get`.